compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -255,44 +255,53 @@ static Value buildReductionStrategyBlockDistributionPart(
   auto x = mlir::gpu::GPUBlockMappingAttr::get(b.getContext(),
                                                ::mlir::gpu::Blocks::DimX);
   // Step 2. First level of tiling + fusion parallelizes to blocks using
-  // `tileSizes`.
+  // `tileSizes`. If the fusion root was the reduction op, update it to be the
+  // combiner op. Otherwise, fuse the combiner op into root.
+  SmallVector<Value> opsHToFuse({originalFillH, splitFillH, splitLinalgH});
+  if (reductionH == fusionRootH) {
+    fusionRootH = combinerH;
+  } else {
+    opsHToFuse.push_back(combinerH);
+  }
   buildTFDWithTileSizes<TileToForeachThreadAndWorkgroupCountRegionOp>(
-      b,
-      /*rootH=*/combinerH,
-      /*opsHToFuse=*/{originalFillH, splitFillH, splitLinalgH},
-      tileSizes0Generic, b.getArrayAttr({x}));
+      b, fusionRootH, opsHToFuse, tileSizes0Generic, b.getArrayAttr({x}));
 
   return variantH;
 }
 
 static Value buildReductionStrategyThreadDistributionPart(
     ImplicitLocOpBuilder &b, Value variantH, ArrayRef<int64_t> tileSizes1Fill,
-    ArrayRef<int64_t> tileSizes1Generic) {
+    ArrayRef<int64_t> tileSizes1Generic, bool hasTrailingEltwise) {
   // TODO: Relying on ordering is brittle, harden this.
-  auto [fill2dH, parParRedH, fill1dH, parRedH] = matchAndUnpack<4>(
-      b, variantH,
-      ArrayRef<StringRef>{linalg::GenericOp::getOperationName(),
-                          linalg::FillOp::getOperationName()});
+  Value matchedH = b.create<MatchOp>(
+      variantH, ArrayRef<StringRef>{linalg::GenericOp::getOperationName(),
+                                    linalg::FillOp::getOperationName()});
+  auto split = b.create<SplitHandlesOp>(
+      matchedH, /*numResultHandles=*/hasTrailingEltwise ? 5 : 4);
+  Value firstFusionRootH = split.getResults()[1];
```

Comment:
A long time ago I used to have something like.

```
Value v1, v2, maybeV3;
bind</*NumRequiredLeading=*/2, /*NumOptional=*/1,/*NumRequiredTrailing=*/0>(split.getResults(), SmallVector<Value*>{&v1, &v2, &maybeV3});
```

with bind doing the right thing.

Something low-tech like this could improve your naming and you wouldn't have to create these arcane unnamed:
```
SmallVector<Value> xxxFusionGroupHs
```
?

Reply:
As discussed offline, this wouldn't scale beyond very simple optionality case. Specifically, it will not be able to differentiate which of the optionals is present in the optional-required-optional sequence with two elements given. We will rather go for a better connection to C++ matchers.

Reply:
+1